libpod/image/image.go

```diff
@@ -0,0 +1,192 @@
+}
+
+// New creates a new image object where the image could be local
+// or remote
+func New(name string, runtime *libpod.Runtime) (Image, error) {
+	// We don't know if the image is local or not ... check local first
+	newImage := Image{
+		InputName: name,
+		Local:     false,
+		runtime:   runtime,
+	}
+	localImage, err := newImage.getLocalImage()
+	if err == nil {
+		newImage.Local = true
+		newImage.image = localImage
+		return newImage, nil
+	}
+
+	// The image is not local
+	pullNames, err := newImage.createNamesToPull()
+	if err != nil {
+		return newImage, err
+	}
+	if len(pullNames) == 0 {
+		return newImage, errors.Errorf("unable to pull %s", newImage.InputName)
+	}
+	var writer io.Writer
+	writer = os.Stderr
+	for _, p := range pullNames {
+		_, err := newImage.pull(p, writer, runtime)
+		if err == nil {
+			newImage.InputName = p
+			img, err := newImage.getLocalImage()
+			newImage.image = img
+			return newImage, err
+		}
+	}
+	return newImage, errors.Errorf("unable to find %s", name)
+}
+
+// getLocalImage resolves an unknown input describing an image and
+// returns a storage.Image or an error. It is used by NewFromLocal.
+// TODO: I think this should be a local func only, forcing users to use NewFromLocal
+func (i *Image) getLocalImage() (*storage.Image, error) {
+	imageError := fmt.Sprintf("unable to find '%s' in local storage\n", i.InputName)
+	if i.InputName == "" {
+		return nil, errors.Errorf("input name is blank")
+	}
+	var taggedName string
+	img, err := i.runtime.GetImage(i.InputName)
+	if err == nil {
+		return img, err
+	}
+
+	// container-storage wasn't able to find it in its current form
+	// check if the input name has a tag, and if not, run it through
+	// again
+	decomposedImage, err := decompose(i.InputName)
+	if err != nil {
+		return nil, err
+	}
+	// the inputname isn't tagged, so we assume latest and try again
+	if !decomposedImage.isTagged {
+		taggedName = fmt.Sprintf("%s:latest", i.InputName)
+		img, err = i.runtime.GetImage(taggedName)
+		if err == nil {
+			return img, nil
+		}
+	}
+	hasReg, err := i.hasRegistry()
+	if err != nil {
+		return nil, errors.Wrapf(err, imageError)
+	}
+
+	// if the input name has a registry in it, the image isnt here
+	if hasReg {
+		return nil, errors.Errorf("%s", imageError)
+	}
+
+	// grab all the local images
+	images, err := i.runtime.GetImages(&libpod.ImageFilterParams{})
+	if err != nil {
+		return nil, err
+	}
+
+	// check the repotags of all images for a match
+	repoImage, err := findImageInRepotags(decomposedImage, images)
+	if err == nil {
+		return repoImage, nil
+	}
+
+	return nil, errors.Errorf("%s", imageError)
+}
+
+// hasRegistry returns a bool/err response if the image has a registry in its
+// name
+func (i *Image) hasRegistry() (bool, error) {
+	imgRef, err := reference.Parse(i.InputName)
+	if err != nil {
+		return false, err
+	}
+	registry := reference.Domain(imgRef.(reference.Named))
+	if registry != "" {
+		return true, nil
+	}
+	return false, nil
+}
```

Comment:
Just wondering if it might be useful to change this to getRegistry() and then return the  registry/nil instead of true/false?  I'm not sure if we'd need the value of the registry, but it might be handy.

Reply:
NVM, just ran across decompose() below, we can get the registry that way.